xontrib/vox.py

```diff
@@ -0,0 +1,318 @@
+        """Create a virtual environment in $VIRTUALENV_HOME with python3's ``venv``.
+
+        Parameters
+        ----------
+        name : str
+            Virtual environment name
+        """
+        print('Creating environment...')
+        self.vox.create(name)
+        msg = 'Environment {0!r} created. Activate it with "vox activate {0}".\n'
+        print(msg.format(name))
+
+    def activate_env(self, name):
+        """Activate a virtual environment.
+
+        Parameters
+        ----------
+        name : str
+            Virtual environment name
+        """
+
+        try:
+            self.vox.activate(name)
+        except KeyError:
+            print('This environment doesn\'t exist. Create it with "vox new %s".\n' % name)
+            return None
+        else:
+            print('Activated "%s".\n' % name)
+
+    def deactivate_env(self):
+        """Deactive the active virtual environment."""
+
+        if self.vox.active() is None:
+            print('No environment currently active. Activate one with "vox activate".\n')
+            return None
+        env_name = self.vox.deactivate()
+        print('Deactivated "%s".\n' % env_name)
+
+    def list_envs(self):
+        """List available virtual environments."""
+
+        try:
+            envs = list(self.vox.keys())
+        except PermissionError:
+            print('No permissions on VIRTUALENV_HOME'.format(venv_home))
+            return None
+
+        if not envs:
+            print('No environments available. Create one with "vox new".\n')
+            return None
+
+        print('Available environments:')
+        print('\n'.join(envs))
+
+
+    def remove_envs(self, *names):
+        """Remove virtual environments.
+
+        Parameters
+        ----------
+        names : list
+            list of virtual environment names
+        """
+        for name in names:
+            try:
+                del self.vox[name]
+            except EnvironmentInUse:
+                print('The "%s" environment is currently active. In order to remove it, deactivate it first with "vox deactivate %s".\n' % (name, name),
+                      file=sys.stderr)
+                return
+            else:
+                print('Environment "%s" removed.' % name)
+        print()
+
+    def show_help(self):
+        """Show help."""
+
+        print(self.__doc__, '\n')
+        self.print_commands()
+
+    @staticmethod
+    def print_commands():
+        """Print available vox commands."""
+
+        print("""Available commands:
+    vox new <env>
+        Create new virtual environment in $VIRTUALENV_HOME
+
+    vox activate (workon, enter) <env>
+        Activate virtual environment
+
+    vox deactivate (exit)
+        Deactivate current virtual environment
+
+    vox list (ls)
+        List all available environments
+
+    vox remove (rm, delete, del) <env> <env2> ...
+        Remove virtual environments
+
+    vox help (-h, --help)
+        Show help
+""")
+
+
```

Comment:
It might be a good idea to split this into a `voxapi.py` or `voxhooks.py` file for everything above this line and leave `vox.py` for everything below this line.  Currently all of the names and definitions will be loaded into the xonsh context. for classes like Vox this isn't a huge deal, but we probably don't want to define os, scandir, etc for everyone.

Note that the mpl.py and mplhooks.py in the xontribs dir takes this same dual file strategy to separate out the API from the xontrib interface.